src/translator/quality_estimator.cpp

```diff
@@ -25,37 +54,61 @@ void QualityEstimator::load(const char* ptr, size_t blobSize) {
   intercept_ = (begin += numFeatures_);
-  modelParameters_.resize(numParameters * numFeatures_);
-  //[0,1,2,3]-> stds, means, coefficients, intercept
-  for (int i = 0; i < numFeatures_; i++) {
-    modelParameters_[i + 0 * numFeatures_] = stds_[i];
-    modelParameters_[i + 1 * numFeatures_] = means_[i];
-    modelParameters_[i + 2 * numFeatures_] = coefficients_[i];
-    modelParameters_[i + 3 * numFeatures_] = intercept_[i];
-  }
+
+  modelMatrix_ = buildLogisticModel();
 }
 
-QualityEstimator::SentenceQualityEstimate QualityEstimator::mapBPEToWords(Quality quality, AnnotatedText target,
-                                                                          size_t sentenceIdx) const {
-  const ByteRange& sentenceByteRange = target.sentenceAsByteRange(sentenceIdx);
-  QualityEstimator::SentenceQualityEstimate sentenceQualityScores;
+std::pair<std::vector<ByteRange>, QualityEstimator::ModelFeatures> QualityEstimator::mapBPEToWords(
+    const std::vector< float > &logProbs, const AnnotatedText& target, const size_t sentenceIdx) const {
+  const auto sentenceByteRange = target.sentenceAsByteRange(sentenceIdx);
+  std::vector<ByteRange> wordByteRanges;
+  WordsQualityEstimate wordsQualityEstimate;
   int bpeLen = 1;
   int wordIdx = 0;
   int subwordIdx = wordIdx;
-  float tmp;
-  for (float& subwordScore : quality.word) {
-    ByteRange subword = target.wordAsByteRange(sentenceIdx, subwordIdx);
+
+  ModelFeatures modelFeatures;
+
+  const auto insertNewWord = [&wordByteRanges, &modelFeatures](const ByteRange& subword, const float subwordScore,
+                                                               const int lenSubwords) {
+    modelFeatures.numSubWords.push_back(lenSubwords);
+    modelFeatures.minWordScores.push_back(subwordScore);
+    modelFeatures.wordMeanScores.push_back(subwordScore);
+    wordByteRanges.push_back(subword);
+  };
+
+  const auto argumentGivenWord = [&wordByteRanges, &modelFeatures](const ByteRange& subword, const float subwordScore,
+                                                                   const int lenSubwords, const int wordIndex) {
+    ByteRange& currentWord = wordByteRanges.back();
+    float& currentScores = modelFeatures.wordMeanScores.back();
+    float& minScore = modelFeatures.minWordScores.at(wordIndex - 1);
+    modelFeatures.numSubWords.at(wordIndex - 1) = lenSubwords;
+    currentWord.end = subword.end;
+    // incremental mean
+    currentScores = currentScores + (subwordScore - currentScores) / lenSubwords;
+    if (minScore > subwordScore) {
+      minScore = subwordScore;
+    }
+  };
+
+  float sequence = 0.0;
+
+  for (const float& subwordScore : logProbs) {
+    sequence += subwordScore;
+
+    auto subword = target.wordAsByteRange(sentenceIdx, subwordIdx);
+
     if (subword.begin == sentenceByteRange.end) {  // EOS token
-      insertNewWord(sentenceQualityScores, subword, subwordScore, 1);
+      insertNewWord(subword, subwordScore, 1);
       break;
```

Comment:
Just to notice here that the reason for this `break` is that it is the only way that I found to collect the `EOS` token without `target.text` breaks since it does not have access to it. I know that this is not the ideal from a software engineering point of view so suggestions are welcome here.